include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+
+/**
+ * @brief Charger Interface
+ * @defgroup charger_interface Charger Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+/**
+ * @brief Runtime Dynamic Battery Parameters
+ */
+enum charger_props {
+	/** Indicates if external supply is present for the charger. */
+	/** Value should be of type enum charger_online */
+	CHARGER_PROP_ONLINE = 0,
+	/** Reports whether or not a battery is present. */
+	/** Value should be of type bool*/
+	CHARGER_PROP_PRESENT,
+	/** Represents the charging status of the charger. */
+	/** Value should be of type enum charger_status */
+	CHARGER_PROP_STATUS,
+	/** Reserved to demark end of common charger properties */
+	CHARGER_PROP_COMMON_COUNT,
+	/**
+	 * Reserved to demark downstream custom properties - use this value as the actual value may
+	 * change over future versions of this API
+	 */
+	CHARGER_PROP_CUSTOM_BEGIN = CHARGER_PROP_COMMON_COUNT + 1,
+	/** Reserved to demark end of valid enum properties */
+	CHARGER_PROP_MAX = UINT16_MAX,
+};
+
+/**
+ * @brief External supply states
+ */
+enum charger_online {
+	/** External supply not presented */
+	CHARGER_ONLINE_OFFLINE = 0,
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Charger prop data types
+ */
+enum charger_type {
+	CHARGER_TYPE_ONLINE,
+	CHARGER_TYPE_BOOL,
+	CHARGER_TYPE_STATUS,
+	CHARGER_TYPE_UINT,
+	CHARGER_TYPE_INT,
+};
+
+/**
+ * @brief A representation of a charger property value
+ */
+struct charger_value {
+	enum charger_type type;
+	union {
+		/* For boolean flags */
+		bool b;
+		/* For unsigned integer values */
+		unsigned int uint;
+		/* For integer values or enums */
+		int intval;
+	} value;
+};
```

Comment:
Hi, sorry I'm very late to the game, apologies if it's been discussed already but I can't find a reference.

Do we really need the union? Could we survive with just `int`? Linux gets away with just `int` or `char *`: https://elixir.bootlin.com/linux/latest/source/include/linux/power_supply.h#L218

Reply:
I think @gmarull was really into the idea of strictly checking the value types. I don't mind either way, but I don't want to redo it if it conflicts with his feedback.

Reply:
I have never asked to "strictly check value types", just a wrapper API around such union type, to make things cleaner e.g. to not expose union internals. IIRC the union has been there since the 1st iteration.

Reply:
okay, so if they're all `int`, do you mind?